src/peft/utils/safelora.py

```diff
@@ -0,0 +1,249 @@
+                tensor = f.get_tensor(name)
+            except SafetensorError as exc:
+                # no matching key found, we probably need to remove the base model prefix
+                if self.base_model_prefix:
+                    # remove 1 extra character for "."
+                    name = name[len(self.base_model_prefix) + 1 :]
+                    tensor = f.get_tensor(name)
+                else:
+                    raise exc
+        return tensor
+
+
+def get_aligned_matrix(base_model_path, aligned_model_path, devices, peft_config):
+    """
+    Get projected matrix by following the config (target_modules) from the peft model.
+    The dimensions between the base model's weights and the aligned model's weights should be the same.
+    """
+    sl_align = SafetensorLoader(aligned_model_path)
+    sl_base = SafetensorLoader(base_model_path)
+
+    base = [name for name in sl_base.weight_map.keys() if any(v in name for v in list(peft_config["target_modules"]))]
+    align = [
+        name for name in sl_align.weight_map.keys() if any(v in name for v in list(peft_config["target_modules"]))
+    ]
+    v = []
+    for name_base, name_align in zip(base, align):
+        assert (
+            sl_base.get_tensor(name_base).shape == sl_align.get_tensor(name_align).shape
+        ), "The dimensions of the base model's weight should be the same with the aligned model's weight."
+        vec = sl_base.get_tensor(name_base) - sl_align.get_tensor(name_align)
+        vec = vec.to(devices)
+        if devices == "cpu":
+            vec = vec.to(torch.float32)
+        vec = torch.mm(vec, vec.t()) / torch.norm(vec)
+        v.append((vec).detach().cpu())
+    return v
+
+
+def project_weights(configs, peft_weights, v):
+    ori_peft_weights = copy.deepcopy(peft_weights)
+    names_A = [name for name in peft_weights.keys() if "lora_A" in name]
+    names_B = [name for name in peft_weights.keys() if "lora_B" in name]
+    idx = 0
+    i = 0
+    dis = []
+    cos_total = []
+    for name_A, name_B in zip(names_A, names_B):
+        A = ori_peft_weights[name_A]
+        if configs.devices != "cpu":
+            P = v[idx].to(torch.bfloat16).to(configs.devices)
+        else:
+            P = v[idx].to("cpu")
+        W = torch.mm(P, ori_peft_weights[name_B])
+        fW = torch.mm(W, A)
+        ori = torch.mm(ori_peft_weights[name_B], A)
+
+        cos = numpy.round(torch.nn.functional.cosine_similarity(fW.reshape(1, -1), ori.reshape(1, -1)).item(), 5)
+        cos_total.append(cos)
+        if cos <= configs.threshold:
+            i += 1
+            peft_weights[name_B] = W
+        else:
+            peft_weights[name_B] = ori_peft_weights[name_B]
+
+        dist = 1 / (1 + torch.norm(peft_weights[name_B].reshape(1, -1) - W.reshape(1, -1)))
+
+        dis.append(dist.item())
+        idx += 1
+    return peft_weights, cos_total
+
+
+def apply_safelora(configs):
+    """
+    ===================================================
```

Comment:
Let's remove the `=`. Also, the first line should be a description of what this function does. You can also include a link to the paper.

Regarding the syntax for examples, check for instance here: https://github.com/huggingface/peft/blob/8eeae0a63f28a80ef42b6bfc977c34ffc0b650ff/src/peft/peft_model.py#L1404-L1429